### DISCUSSION TITLE
series_analysis generating NetCDF dimension error for matched pairs produced by grid_stat

### DISCUSSION THREAD

**Discussion by erinthead** [dtcenter/METplus#1773] [File I/O] [answered]
Hi,

I'm trying to run series_analysis on WCOSS2 on some grid_stat output -- specifically, the matched pairs NetCDF output generated by grid_stat. However, I'm unable to get a run to complete successfully. My logs show the following error:

> terminate called after throwing an instance of 'netCDF::exceptions::NcBadDim'
>   what():  NetCDF: Invalid dimension ID or name
> file: ../../cxx4/ncDim.cpp  line:109

This error always occurs after it has generated a few GSL_RNG_SEEDs. In my regular output log, it indicates that it has found the data it's looking for before it throws the error:

> DEBUG 2: Found data for FCST_dp_Z2_FULL/Z2 in file: /lfs/h1/mdl/ptmp/erin.thead/series_analysis/dp_202206_0_6/grid_stat_060000L_20220601_060000V_pairs.nc
> DEBUG 3: Setting the observation valid time search criteria using the forecast valid time of 20220601_060000.
> DEBUG 3: Searching file /lfs/h1/mdl/ptmp/erin.thead/series_analysis/dp_202206_0_6/grid_stat_060000L_20220601_060000V_pairs.nc
> DEBUG 2: Found data for OBS_dp_Z2_FULL/Z2 in file: /lfs/h1/mdl/ptmp/erin.thead/series_analysis/dp_202206_0_6/grid_stat_060000L_20220601_060000V_pairs.nc
> DEBUG 3: Grid Definition: Projection: Lambert Conformal Nx: 2345 Ny: 1597 Lat_LL: 19.229 Lon_LL: 126.277 Lon_orient: 95.000 Alpha: 6509.082 Cone: 0.423 Bx: 1288.0039 By: 5483.6648

The headers of these input NetCDFs look like this (an example; I am giving series_analysis a month of such data):

> netcdf grid_stat_060000L_20220601_060000V_pairs {
> dimensions:
>         lat = 1597 ;
>         lon = 2345 ;
> variables:
>         float lat(lat, lon) ;
>                 lat:long_name = "latitude" ;
>                 lat:units = "degrees_north" ;
>                 lat:standard_name = "latitude" ;
>         float lon(lat, lon) ;
>                 lon:long_name = "longitude" ;
>                 lon:units = "degrees_east" ;
>                 lon:standard_name = "longitude" ;
>         float FCST_dp_Z2_FULL(lat, lon) ;
>                 FCST_dp_Z2_FULL:name = "FCST_dp_Z2_FULL" ;
>                 FCST_dp_Z2_FULL:long_name = "Dew point temperature at Z2" ;
>                 FCST_dp_Z2_FULL:level = "Z2" ;
>                 FCST_dp_Z2_FULL:units = "K" ;
>                 FCST_dp_Z2_FULL:init_time = "20220601_000000" ;
>                 FCST_dp_Z2_FULL:init_time_ut = "1654041600" ;
>                 FCST_dp_Z2_FULL:valid_time = "20220601_060000" ;
>                 FCST_dp_Z2_FULL:valid_time_ut = "1654063200" ;
>                 FCST_dp_Z2_FULL:accum_time = "NA" ;
>                 FCST_dp_Z2_FULL:accum_time_sec = -9999 ;
>                 FCST_dp_Z2_FULL:_FillValue = -9999.f ;
>                 FCST_dp_Z2_FULL:desc = "NA" ;
>                 FCST_dp_Z2_FULL:masking_region = "FULL" ;
>                 FCST_dp_Z2_FULL:smoothing_method = "NEAREST" ;
>                 FCST_dp_Z2_FULL:smoothing_neighborhood = 1 ;
>         float OBS_dp_Z2_FULL(lat, lon) ;
>                 OBS_dp_Z2_FULL:name = "OBS_dp_Z2_FULL" ;
>                 OBS_dp_Z2_FULL:long_name = "Dew point temperature at Z2" ;
>                 OBS_dp_Z2_FULL:level = "Z2" ;
>                 OBS_dp_Z2_FULL:units = "K" ;
>                 OBS_dp_Z2_FULL:init_time = "20220601_060000" ;
>                 OBS_dp_Z2_FULL:init_time_ut = "1654063200" ;
>                 OBS_dp_Z2_FULL:valid_time = "20220601_060000" ;
>                 OBS_dp_Z2_FULL:valid_time_ut = "1654063200" ;
>                 OBS_dp_Z2_FULL:accum_time = "NA" ;
>                 OBS_dp_Z2_FULL:accum_time_sec = -9999 ;
>                 OBS_dp_Z2_FULL:_FillValue = -9999.f ;
>                 OBS_dp_Z2_FULL:desc = "NA" ;
>                 OBS_dp_Z2_FULL:masking_region = "FULL" ;
>                 OBS_dp_Z2_FULL:smoothing_method = "NEAREST" ;
>                 OBS_dp_Z2_FULL:smoothing_neighborhood = 1 ;

My series_analysis config file has this for the forecast and observation fields:

> fcst = {
> 
>    field = [
>       {
>         name  = "FCST_${FIELD}_${LEVEL}_FULL";
>         // name = "${FIELD}";
>         level = "${LEVEL}";
>       }
>    ];
> 
> }
> obs = {
>    field = [
>       {
>         name  = "OBS_${FIELD}_${LEVEL}_FULL";
>         //name = "${FIELD}";
>         level = "${LEVEL}";
>       }
>    ];
> }

I've confirmed that the environment variables I am passing in have the correct values. I've also confirmed that the matched pairs NetCDF contains values and that they are reasonable.  The grid_stat run appears to complete correctly and generate valid output, but for some reason series_analysis can't handle this output as input.

I'd also like to note that if I change the series_analysis runtime arguments to use separate forecast and obs files (which I have created with regrid_data_plane for use in grid_stat), and change the config file in the obvious way (comment/uncomment the appropriate lines for field.name), I get that same NetCDF dimension error.

The grid_stat run is taking place in the same job as the series_analysis run on WCOSS2, so there would be no differences in modules loaded.  I don't understand what could be going wrong with this and I'm hoping it's just a configuration file issue I've overlooked.  Can anyone help?

**Comment by CPKalb** [comment]
Hi:

It looks like you are using the same value of ${LEVEL} in the variable name and level.  For a netCDF file, the level is usually specified as something like the following: "(* , *)", where the stars are for the 2 dimensions.  If there were more dimensions, then this specification would be different.  So, for your FCST_dp_Z2_FULL variable, you would want something like this in the configuration file:
fcst = {
field = [
{
name = "FCST_dp_Z2_FULL";
level = "( * , * )";
}
];

Without seeing the values of your environment variables, it's hard to know for sure, but I'm guessing you may need to set the level to   "(* , *)", rather than setting it as "Z2".  Does making this change have any impact on the error?

As a side note, you don't need the spaces in the level value.  However, I couldn't get them to display properly in this thread without adding them.

Thanks,
Christina

**Reply by erinthead** [reply]
Hi Christina,

Thanks for the suggestion. Unfortunately I have tried that already in the config files and it too produces the same error.

**Reply by CPKalb** [reply]
Hi:

Would you be able to send us a sample file as well as your series analysis configuration file so that we can take a closer look?  The configuration file could probably be uploaded here, and instructions for sending data can be found [here](https://github.com/dtcenter/METplus/discussions/954).  

Christina

**Reply by erinthead** [reply]
The files should be there. The environment variables I invoke are set in the run script, but here they are:

export FORECAST=NDFD
export FIELD=dp
export METVER=10.0.1
export LEVEL=Z2
export TMP=/lfs/h1/mdl/ptmp/$USER

I am calling series_analysis with -both set to the file list (there are many files like the netCDF I attached, all generated from grid_stat, but the error occurs with every one of them that it attempts to read).

**Reply by CPKalb** [reply]
Hi:

To double check, are you running MET version 10.0.1?

Christina

**Reply by CPKalb** [reply]
Also, I don't have access to your mask file.  Would you be able to send that over?

Thanks,
Christina

**Reply by erinthead** [reply]
Hi Christina,

Sorry for the delay -- holiday weekend. The mask file should be in the directory as well. Please let me know if you require anything else.

**Reply by CPKalb** [reply]
Thanks.  The uploaded mask has a file name of ndfd_co_ws.nc, while the one in the file says blend_co_ws.nc.  Are these the same mask?

Christina

**Comment by erinthead** [comment]
Hi Christina,

I don't know; I did not develop these mask files. I have uploaded the blend one as well. It's the same size as the ndfd one, but it's possible they contain different geographic regions in terms of actual data points. Best to use the one that is referenced in the file.

**Comment by CPKalb** [comment, marked as answer]
Hi:

It's actually the mask file that's causing the error.  When I comment out the mask file, or use ndfd_co_ws.nc, then series_analysis runs without error.   I need to dig a little bit more to figure out what needs to be changed to get this mask to work in MET, but I'm guessing that the dimension names need to be changed to say lat and lon, rather than latitude and longitude.

It does looks like ndfd_co_ws.nc is the same domain, and that mask does not produce an error in MET.

Christina

**Reply by erinthead** [reply]
Hi Christina,

Thanks so much for looking into that. I've let my team know about the issue with that mask file. I do need to use the blend domain, so I'm going to try to rename the dimensions within the file. If need be, my team may also need to recreate the mask. I'm going to mark this as an answer and I'll return if we continue to have problems after trying those things. Thanks again!

Erin

**Reply by CPKalb** [reply]
Hi:

Both mask files state that they were created by regrid_data_plane in the netCDF attributes.  Usually, masks in MET are a 0/1 binary field that are created using gen_vx_mask.  I have not been able to do a ncview on the blend mask to see what the data looks like.  However, ndfd_co_ws.nc is a data file rather than a 0/1 mask.  I suspect that gen_vx_mask will need to be run to get a 0/1 binary.

To run gen_vx_mask usually requires a data file and then a specification of the domain to retain.  This domain could be a list of points bounding the region, or something like elevation > 1000ft (if you wanted to retain data based on something like an elevation).  It's best to start a new discussion if you have questions on gen_vx_mask.

Thanks,
Christina
